### PR TITLE
[CI] Prune unused archs from libnccl

### PR DESCRIPTION
**Pull request by hcho3**
This further reduces the size of the compiled binary.

### CODE REVIEW

**Review by trivialfis — CHANGES_REQUESTED** [review]
Thank you for working on this. I think this needs to be a CI specific change. People might not want XGBoost to prune the nccl on their system.

File: CMakeLists.txt [hunk 1]
```diff
@@ -141,6 +141,10 @@ if (USE_CUDA)
   if ((${CMAKE_CUDA_COMPILER_VERSION} VERSION_GREATER_EQUAL 11.4) AND (NOT BUILD_WITH_CUDA_CUB))
     set(BUILD_WITH_CUDA_CUB ON)
   endif ()
+
+  if (USE_NCCL)
+    add_custom_target(PruneLibNccl nvprune ${GEN_CODE} /usr/lib64/libnccl_static.a -o libnccl_static.a)
```
Comment (RAMitchell) [inline CMakeLists.txt:146]: Should this path be hardcoded?

**Review by RAMitchell — COMMENTED** [review]
Agree with Jiaming that users self installing probably don't need this, and it could easily break their install process.

Is it possible to create the option PRUNE_NCCL which performs the prune then modifies the nccl library path variable in find_nccl.

**Comment by hcho3** [comment]
> People might not want XGBoost to prune the nccl on their system.

The custom target is entirely optional, as it's not part of the `all` target. So you need to explicitly opt in: `ninja PruneLibNccl`

**Comment by hcho3** [comment]
> Is it possible to create the option PRUNE_NCCL which performs the prune then modifies the nccl library path variable in find_nccl.

This is hard, as you'd need to modify the nccl path as a result of building a target.

**Comment by hcho3** [comment]
@trivialfis @RAMitchell I revised this pull request so that the changes are restricted to scripts in the CI directory `tests/ci_build`. Can you review again?

**Comment by RAMitchell** [comment]
Thanks. I was curious if it can be done in pure cmake with a bit less hacking, here is my attempt: #8183. Not sure if it's any better.

**Comment by RAMitchell** [comment]
Let's go with this version I think.

**Review by RAMitchell — APPROVED** [review]

**Review by trivialfis — APPROVED** [review]

**Pull request merged**